kola/tests/crio/crio.go

```diff
@@ -0,0 +1,374 @@
+	"github.com/coreos/mantle/lang/worker"
+	"github.com/coreos/mantle/platform"
+)
+
+// crioArguments abstracts arguments used within a crio json config
+type crioArguments []string
+
+// simplifiedCrioInfo represents the results from crio info
+type simplifiedCrioInfo struct {
+	StorageDriver string `json:"storage_driver"`
+	StorageRoot   string `json:"storage_root"`
+	CgroupDriver  string `json:"cgroup_driver"`
+}
+
+// crioPodTemplate is a simple string template required for creating a pod in crio
+// It takes two strings: the name (which will be expanded) and the generated image name
+var crioPodTemplate = `{
+	"metadata": {
+		"name": "rhcos-crio-pod-%s",
+		"namespace": "redhat.test.crio"
+	},
+	"image": {
+			"image": "localhost/%s:latest"
+	},
+	"args": [],
+	"readonly_rootfs": false,
+	"log_path": "",
+	"stdin": false,
+	"stdin_once": false,
+	"tty": true,
+	"linux": {
+			"resources": {
+					"memory_limit_in_bytes": 209715200,
+					"cpu_period": 10000,
+					"cpu_quota": 20000,
+					"cpu_shares": 512,
+					"oom_score_adj": 30,
+					"cpuset_cpus": "0",
+					"cpuset_mems": "0"
+			},
+			"cgroup_parent": "Burstable-pod-123.slice",
+			"security_context": {
+					"namespace_options": {
+							"pid": 1
+					},
+					"capabilities": {
+							"add_capabilities": [
+								"sys_admin"
+							]
+					}
+			}
+	}
+}`
+
+// crioContainerTemplate is a simple string template required for running a container
+// It takes three strings: the name (which will be expanded), the image, and the argument to run
+var crioContainerTemplate = `{
+	"metadata": {
+		"name": "rhcos-crio-container-%s",
+		"attempt": 1
+	},
+	"image": {
+		"image": "docker.io/library/%s"
+	},
+	"command": [
+		"%s"
+	],
+	"args": [],
+	"working_dir": "/",
+	"envs": [
+		{
+			"key": "PATH",
+			"value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+		},
+		{
+			"key": "TERM",
+			"value": "xterm"
+		}
+	],
+	"labels": {
+		"type": "small",
+		"batch": "no"
+	},
+	"annotations": {
+		"daemon": "crio"
+	},
+	"privileged": true,
+	"log_path": "",
+	"stdin": false,
+	"stdin_once": false,
+	"tty": false,
+	"linux": {
+		"resources": {
+			"cpu_period": 10000,
+			"cpu_quota": 20000,
+			"cpu_shares": 512,
+			"oom_score_adj": 30,
+			"memory_limit_in_bytes": 268435456
+		},
+		"security_context": {
+			"readonly_rootfs": false,
+			"selinux_options": {
+				"user": "system_u",
+				"role": "system_r",
+				"type": "svirt_lxc_net_t",
+				"level": "s0:c4,c5"
+			},
+			"capabilities": {
+				"add_capabilities": [
+					"setuid",
+					"setgid"
+				],
+				"drop_capabilities": [
+				]
+			}
+		}
+	}
+}`
+
+// init runs when the package is imported and takes care of registering tests
+func init() {
+	register.Register(&register.Test{
+		Run:              crioBaseTests,
+		ClusterSize:      1,
+		Name:             `crio.base`,
+		ExcludePlatforms: []string{"qemu"}, // crio pods require fetching a kubernetes pause image
```

Comment:
`crio` itself pulls down the pause image as part of how it works. The image is a normal part of kubernetes/openshift. Since `crio` is built specifically for kubernetes/openshift it runs with the requirement to pull the pause before running a container/pod.

Reply:
Hmm, I guess it just doesn't log that pull then?

I can see the other images getting referenced in the build logs.

```
--- PASS: crio.base (165.85s)
    --- PASS: crio.base/crio-info (1.52s)
    --- PASS: crio.base/networks-reliably (44.53s)
            cluster.go:104: STEP 1: FROM scratch
            cluster.go:104: STEP 2: COPY . /
            cluster.go:104: STEP 3: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]docker.io/library/ping:latest
            cluster.go:104: Getting image source signatures
            cluster.go:104: Copying blob sha256:bdd3c86d8bac096cb6fd677ec9a8df0a8fe36d1c564524445dbddb31e74bb0fc
 2.94 MiB / 2.94 MiB  0s04: 
            cluster.go:104: Copying config sha256:3e8ec409a8bd4f20eab6d09a39c2ce4c3184593f4e039cadca2d06e32791c214
 355 B / 355 B  0sr.go:104: 
            cluster.go:104: Writing manifest to image destination
            cluster.go:104: Storing signatures
--- PASS: crio.network (122.80s)
        crio.go:229: creating ncat containers
        cluster.go:104: STEP 1: FROM scratch
        cluster.go:104: STEP 2: COPY . /
        cluster.go:104: STEP 3: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]docker.io/library/ncat:latest
        cluster.go:104: Getting image source signatures
        cluster.go:104: Copying blob sha256:c25a8e68e58eb4c7031f747a2f9aa5a3e08f667a0c31209444c1e466613d8cee
 4.23 MiB / 4.23 MiB  0s
        cluster.go:104: Copying config sha256:fac71e83e204a3358d8c2e53296884233933a9d2093e440419cebef1ffd885a5
 355 B / 355 B  0s:104: 
        cluster.go:104: Writing manifest to image destination
        cluster.go:104: Storing signatures
        cluster.go:104: STEP 1: FROM scratch
        cluster.go:104: STEP 2: COPY . /
        cluster.go:104: STEP 3: COMMIT containers-storage:[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]docker.io/library/ncat:latest
        cluster.go:104: Getting image source signatures
        cluster.go:104: Copying blob sha256:2f63e66371da2d0748d0896b2c5622dcd06f648da56e8a6c5e3632ce844f8a1a
 4.23 MiB / 4.23 MiB  0s
        cluster.go:104: Copying config sha256:cf4d92b09a3c19f4ca768fac2f442732cb2008442ea3352255a273889c42bc80
 353 B / 353 B  0s:104: 
        cluster.go:104: Writing manifest to image destination
        cluster.go:104: Storing signatures
```

Reply:
The logged images are actually built from binaries on the host, to make testing images. The pull for the pause image happens behind the scenes during the first pod run `sudo crictl runp %s`. I recall seeing it the test logs somewhere.